File: Raven.Abstractions/OAuth/AbstractAuthenticator.cs
using System;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using Raven.Abstractions.Connection;

namespace Raven.Abstractions.OAuth
{
	public abstract class AbstractAuthenticator
	{
		protected string CurrentOauthToken { get; set; }

		public virtual void ConfigureRequest(object sender, WebRequestEventArgs e)
		{
			SetAuthorization(e);
		}

		protected void SetAuthorization(WebRequestEventArgs e)
		{
			if (string.IsNullOrEmpty(CurrentOauthToken))
				return;

			if (e.Client != null)
			{
				SetAuthorization(e.Client);
			}

#if !SILVERLIGHT && !NETFX_CORE
			if (e.Request != null)
				SetHeader(e.Request.Headers, "Authorization", CurrentOauthToken);
#endif
		}

		protected void SetAuthorization(HttpClient e)
		{
			if (string.IsNullOrEmpty(CurrentOauthToken))
				return;

			try
			{
				e.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", CurrentOauthToken);
			}
			catch (Exception ex)
			{
				throw new InvalidOperationException(string.Format("Could not set the Authorization to the value 'Bearer {0}'", CurrentOauthToken), ex);
			}
		}

#if !SILVERLIGHT && !NETFX_CORE
		protected static void SetHeader(WebHeaderCollection headers, string key, string value)
		{
			try
			{
				headers[key] = value;
			}
			catch (Exception e)
			{
				throw new InvalidOperationException("Could not set '" + key + "' = '" + value + "'", e);
			}
		}

		public abstract Action<HttpWebRequest> DoOAuthRequest(string oauthSource, string apiKey);
#endif
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/oauth/AbstractAuthenticator.java
package net.ravendb.abstractions.oauth;

import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.connection.WebRequestEventArgs;

import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpRequest;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpUriRequest;


public abstract class AbstractAuthenticator {

  protected String currentOauthToken;

  public void configureRequest(Object sender, WebRequestEventArgs e) {
    if (StringUtils.isEmpty(currentOauthToken)) {
      return;
    }
    setHeader(e.getRequest(), "Authorization", currentOauthToken);
  }

  protected void setAuthorization(HttpClient httpClient) {
    if (StringUtils.isEmpty(currentOauthToken)) {
      return;
    }
  }

  protected static void setHeader(HttpUriRequest request, String key, String value) {
    request.setHeader(key, value);
  }

  public abstract Action1<HttpRequest> doOAuthRequest(String oauthSource, String apiKey);

}
